castervoice/apps/griddouglas.py

```diff
@@ -13,47 +13,105 @@
+    s.move_mouse(int(x), int(y))
     int_a = int(action)
+    if (int_a == 0) | (int_a == 1):
+        s.kill()
+        grids.wait_for_death(settings.DOUGLAS_TITLE)
     if int_a == 0:
-        Playback([(["mouse", "left", "click"], 0.0)]).execute()
+        Mouse("left").execute()
     elif int_a == 1:
-        Playback([(["mouse", "right", "click"], 0.0)]).execute()
+        Mouse("right").execute()
+
+def send_input_select(x1, y1, x2, y2, nexus):
+    s = nexus.comm.get_com("grids")
+    s.move_mouse(int(x1), int(y1))
+    _x1, _y1 = win32api.GetCursorPos()
+    s.move_mouse(int(x2), int(y2))
+    _x2, _y2 = win32api.GetCursorPos()
+    s.kill()
+    grids.wait_for_death(settings.DOUGLAS_TITLE)
+    drag_from_to(_x1,_y1,_x2,_y2)
+
+def send_input_select_short(x1, y1, x2, nexus):
+    send_input_select(x1, y1, x2, y1, nexus)
 
+def drag_from_to(x1, y1, x2, y2):
+    win32api.SetCursorPos((x1,y1))
+    time.sleep(0.1)
+    win32api.mouse_event(win32con.MOUSEEVENTF_LEFTDOWN,0,0)
+    win32api.SetCursorPos((x2,y2))
+    time.sleep(0.1)
+    win32api.mouse_event(win32con.MOUSEEVENTF_LEFTUP,0,0)
+
+x1 = None
+x2 = None
+y1 = None
+y2 = None
+
+def store_point(point):
+    global x1, y1, x2, y2
+    int_p = int(point)
+    if int_p == 1:
+        x1, y1 = win32api.GetCursorPos()
+    elif int_p == 2:
+        x2, y2 = win32api.GetCursorPos()
+    else:
+        raise Exception("Illegal point specified")
+
+def select_text(nexus):
+    global x1, y1, x2, y2
+    s = nexus.comm.get_com("grids")
+    s.kill()
+    grids.wait_for_death(settings.DOUGLAS_TITLE)
+    drag_from_to(x1,y1,x2,y2)
 
 class GridControlRule(MergeRule):
 
     mapping = {
-        "<n> [by] <n2> [<action>]":
+        "<x> [by] <y> [<action>]":
             R(Function(send_input, nexus=_NEXUS), rdescript="Douglas Grid: Action"),
+        "<x1> [by] <y1> select <x2> [by] <y2>":
+            R(Function(send_input_select, nexus=_NEXUS), rdescript="Douglas Grid: Select (long version)"),
+        "<x1> [by] <y1> select <x2>":
+            R(Function(send_input_select_short, nexus=_NEXUS), rdescript="Douglas Grid: Select (short version)"),
+        "point <point>":
+            R(Function(store_point), rdescript="Douglas Grid: Stor point"),
+        "select":
+            R(Function(select_text, nexus=_NEXUS), rdescript="Douglas Grid: Select (point version)"),
         "exit | escape | cancel":
-            R(Function(kill, nexus=_NEXUS), rdescript="Exit Douglas Grid"),
+            R(Function(kill, nexus=_NEXUS), rdescript="Douglas Grid: Exit"),
     }
     extras = [
-        IntegerRefST("n", 0, 300),
-        IntegerRefST("n2", 0, 300),
+        IntegerRefST("x", 0, 300),
+        IntegerRefST("y", 0, 300),
+        IntegerRefST("x1", 0, 300),
+        IntegerRefST("y1", 0, 300),
+        IntegerRefST("x2", 0, 300),
+        IntegerRefST("y2", 0, 300),
         Choice("action", {
             "kick": 0,
             "psychic": 1,
+            "move": 2,
```

Comment:
"hover"? Does caster have a standard spec for other hover/move commands?

Reply:
"light" https://github.com/dictation-toolbox/Caster/blob/c5074e2a8b7de348cbfa80a4686e7e2d5db2ed9a/caster/apps/gridlegion.py#L73

Reply:
light in legion highlights the word.